components/omega/doc/design/MachEnv.md

```diff
@@ -0,0 +1,263 @@
+
+When running as part of a coupled system, OMEGA will need to define a default
+communicator based on a parent communicator sent by the calling routine
+(coupler or parent model).
+
+### 2.3 Requirement: Define MPI layouts
+
+Each MPI rank will need to know its own rank id, number of ranks and
+define a master rank. 
+
+### 2.4 Requirement: YAKL initialization
+
+Since we are using YAKL for kernel launching and array types, we
+will need to initialize YAKL in standalone mode. It may also be
+needed for coupled simulations.
+
+### 2.5 Requirement: Vector blocking size defined at compile time
+
+To achieve the best CPU performance, especially when using
+GPU-friendly loop forms, it is useful to explicitly size inner
+loops based on a compile-time length (chunk size) that is a multiple
+of the vector length. 
+
+### 2.6 Desired: Set alternative master task
+
+While rank 0 is typically used as a master task, it is sometimes desirable
+to assign a different rank as master to avoid overloading rank 0, especially
+when running in coupled mode with other components also using the same rank.
+
+### 2.7 Desired: Multiple environments
+
+In OMEGA, we may wish to run sub-components on different partitions. For
+example, we might want to rearrange the communication-dominated
+barotropic solve to run on fewer nodes or within a single node. We will
+need to be able to create new environments based on a subset of an 
+existing environment.
+
+### 2.8 Desired: Define threading parameters
+
+If OpenMP threading is enabled for CPU, it may be useful to also define
+similar master threads and thread layouts to enable some task parallelism
+within an MPI rank.
+
+### 2.9 Desired: Define other machine parameters
+
+In the future, it may be useful to define other machine parameters, like
+the specific configuration of cores, accelerators and other devices, to
+manage task assignments on hybrid nodes. It should be easy to modify
+this class to add additional information.
+
+## 3 Algorithmic Formulation
+
+No algorithms are needed beyond what is provided by standard MPI, OpenMP
+libraries.
+
+## 4 Design
+
+In general, this is a simple class to hold information for later
+retrieval.
+
+### 4.1 Data types and parameters
+
+#### 4.1.1 Parameters 
+
+For improved vector performance on CPU and perhaps match thread block
+sizes on GPU, we wish to block the inner loops with a compile-time
+parameter. We define a CPP parameter: `OMEGA_VECTOR_SIZE`. This is 
+dependent on the machine and on whether GPU acceleration is enabled.
+It will typically take on values like 16, 32, 64 for CPU-only builds
+and 1 for GPU builds to maximize parallelism.
+
+#### 4.1.2 Class/structs/data types
+
+There will be a simple class MachEnv. We use a class here rather than
+a struct so that we can make members private to prevent overwriting 
+these variables.
+
+    class MachEnv {
+
+       private:
+          int mComm;      ///< MPI communicator for this environment
```

Comment:
Probably need a color attribute or an abstraction over it to represent sub-communicator category.
https://www.open-mpi.org/doc/v4.1/man3/MPI_Comm_split.3.php

Reply:
I do have the possibility of a color abstraction for the domain decomposition and the separation of owned/halo. We could do the same at this level or push that abstraction down. For now, I think I'll leave this as is.